src/librustc/infer/error_reporting/anon_anon_conflict.rs

```diff
@@ -0,0 +1,197 @@
+                (&ret_ty.sty, self.tcx.hir.get(node_id)) {
+                if let hir::ItemFn(ref fndecl, _, _, _, _, _) = it.node {
+                    fndecl
+                        .inputs
+                        .iter()
+                        .filter_map(|arg| {
+                            let mut nested_visitor = FindNestedTypeVisitor {
+                                infcx: &self,
+                                hir_map: &self.tcx.hir,
+                                bound_region: *br,
+                                found_type: None,
+                            };
+                            nested_visitor.visit_ty(&**arg);
+                            if nested_visitor.found_type.is_some() {
+                                nested_visitor.found_type
+                            } else {
+                                None
+                            }
+                        })
+                        .next()
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn try_report_anon_anon_conflict(&self, error: &RegionResolutionError<'tcx>) -> bool {
+
+        let (span, sub, sup) = match *error {
+            ConcreteFailure(ref origin, sub, sup) => (origin.span(), sub, sup),
+            _ => return false, // inapplicable
+        };
+
+        // Determine whether the sub and sup consist of both anonymous (elided) regions.
+        if let (Some(br1), Some(br2)) =
+            (self.is_anonymous_region(sup), self.is_anonymous_region(sub)) {
+            if let (Some(ty1), Some(ty2)) =
+                (self.find_anon_type(sup, &br1), self.find_anon_type(sub, &br2)) {
+                if let (Some((anon_arg1, _, _, _)), Some((anon_arg2, _, _, _))) =
+                    (self.find_arg_with_anonymous_region(sup, sup),
+                     self.find_arg_with_anonymous_region(sub, sub)) {
+
+                    let span_label_var1 = if let Some(simple_name) = anon_arg1.pat.simple_name() {
+                        format!(" from `{}` ", simple_name)
+                    } else {
+                        format!(" ")
+                    };
+
+                    let span_label_var2 = if let Some(simple_name) = anon_arg2.pat.simple_name() {
+                        format!(" into `{}` ", simple_name)
+                    } else {
+                        format!(" ")
+                    };
+
+                    struct_span_err!(self.tcx.sess, span, E0623, "lifetime mismatch")
+                        .span_label(ty1.span,
+                                    format!("these references must have the same lifetime"))
+                        .span_label(ty2.span, format!(""))
+                        .span_label(span,
+                                    format!("data{}flows{}here", span_label_var1, span_label_var2))
+                        .emit();
+                    return true;
```

Comment:
I would also try to "un-nest" here, so that you have `return false` bailing early, and just focus on the main path.